pkg/webhook/resources/setting/validator.go

```diff
@@ -709,6 +723,159 @@ func (v *settingValidator) validateDeleteStorageNetwork(setting *v1beta1.Setting
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pvc := range pvcs {
+		if pvc.Spec.VolumeName == "" {
+			continue
+		}
+
+		prometheusVolumes = append(prometheusVolumes, pvc.Spec.VolumeName)
+	}
+
+	return prometheusVolumes, nil
+}
+
+func (v *settingValidator) getAlertManagerVolumes() ([]string, error) {
+	alertmanagerVolumes := make([]string, 0)
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValueAlertManager,
+	}
+
+	pvcs, err := v.pvcCache.List(util.CattleMonitoringSystemNamespace, sets.AsSelector())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pvc := range pvcs {
+		if pvc.Spec.VolumeName == "" {
+			continue
+		}
+
+		alertmanagerVolumes = append(alertmanagerVolumes, pvc.Spec.VolumeName)
+	}
+
+	return alertmanagerVolumes, nil
+}
+
+func (v *settingValidator) getGrafanaVolume() (string, error) {
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValueGrafana,
+	}
+
+	pvcs, err := v.pvcCache.List(util.CattleMonitoringSystemNamespace, sets.AsSelector())
+	if err != nil {
+		return "", err
+	}
+
+	if len(pvcs) == 0 {
+		return "", nil
+	}
+
+	//grafana deployment only has one pvc
+	return pvcs[0].Spec.VolumeName, nil
+}
+
+func (v *settingValidator) getVMImportControllerVolume() (string, error) {
+	sets := labels.Set{
+		labelAppNameKey: labelAppNameValueImportController,
+	}
+
+	pvcs, err := v.pvcCache.List(util.HarvesterSystemNamespaceName, sets.AsSelector())
+	if err != nil {
+		return "", err
+	}
+
+	if len(pvcs) == 0 {
+		return "", nil
+	}
+
+	//VM import contoller deployment only has one pvc
+	return pvcs[0].Spec.VolumeName, nil
+}
+
+func (v *settingValidator) getSystemVolumes() (map[string]struct{}, error) {
+	systemVolumes := make(map[string]struct{})
+
+	prometheusVolumes, err := v.getPrometheusVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, volume := range prometheusVolumes {
+		systemVolumes[volume] = struct{}{}
+	}
+
+	alertmanagerVolumes, err := v.getAlertManagerVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, volume := range alertmanagerVolumes {
+		systemVolumes[volume] = struct{}{}
+	}
+
+	grafanaVolume, err := v.getGrafanaVolume()
+	if err != nil {
+		return nil, err
+	}
+
+	if grafanaVolume != "" {
+		systemVolumes[grafanaVolume] = struct{}{}
+	}
+
+	vmImportControllerVolume, err := v.getVMImportControllerVolume()
+	if err != nil {
+		return nil, err
+	}
+
+	if vmImportControllerVolume != "" {
+		systemVolumes[vmImportControllerVolume] = struct{}{}
+	}
+
+	return systemVolumes, nil
+}
+
+func (v *settingValidator) checkOnlineVolume() error {
+	systemVolumes, err := v.getSystemVolumes()
+	if err != nil {
+		logrus.Errorf("getSystemPVCs err: %v", err)
+		return err
+	}
+
+	for volume := range systemVolumes {
+		logrus.Infof("get getSystem volume %v", volume)
+	}
+
+	volumes, err := v.lhVolumeCache.List(util.LonghornSystemNamespaceName, labels.Everything())
+	if err != nil {
+		logrus.Errorf("list longhorn volumes err %v", err)
+		return err
+	}
```

Comment:
this would be a cumbersome call for a large cluster since all ns volumes will be listed here, would prefer to either filter it by labels or get volumes one by one since there won't be too many `systemVolumes`.

Reply:
It's for finding the online Longhorn volumes other than system volumes, so it seems needed to check all Longhorn volumes. Thanks.

Reply:
I know we are looking for the LH volumes, and since the volume name is identical to the pv name, so instead of listing and looping the entire volume list, getting volume by pv and returning directly when it is attached is preferred from my point of view, thanks.

Reply:
But even instead of looping longhorn volumes, we still need to iterate all pvs and check the status of related longhorm volume. Do I misunderstand? thanks.

Reply:
yes, but not iterate all pvc, only those system volumes that u want to check if I understand the code correctly.

Reply:
Actually, we want the check the status of all non-system volumes here, that's why it requires looping over PVs/Longhorn volumes. However, I tried to iterate Longhorn volumes with pagination, and it works well so far. With pagination, it retried limited numbers of Longhorn volumes at a time, and if there is any non-system volume that meets certain conditions, we can break out of the loop. So we don't have to list all the Longhorn volumes most of the time. I will update this implementation with other fixes,  thanks.

Reply:
I see, if it is about to check all non-system volume, then this LGTM, thanks.